File: ios/Classes/Renderer/AxisRenderer.swift
import Foundation
import UIKit

final class AxisRenderer: ChartLayerRenderer {
  func render(frame: ChartFrame) {
    guard frame.config.showAxis, let priceAxisRect = frame.priceAxisRect, let timeAxisRect = frame.timeAxisRect else {
      return
    }

    let context = frame.context
    context.setStrokeColor(frame.theme.axisColor)
    context.setLineWidth(1)
    context.move(to: CGPoint(x: frame.mainChartRect.maxX, y: frame.mainChartRect.minY))
    context.addLine(to: CGPoint(x: frame.mainChartRect.maxX, y: frame.mainChartRect.maxY))
    context.move(to: CGPoint(x: frame.mainChartRect.minX, y: frame.mainChartRect.maxY))
    context.addLine(to: CGPoint(x: frame.mainChartRect.maxX, y: frame.mainChartRect.maxY))
    context.strokePath()

    guard frame.viewport.hasVisibleRange() else { return }

    UIGraphicsPushContext(context)
    defer { UIGraphicsPopContext() }

    let labelCount = 4
    for step in 0...labelCount {
      let ratio = CGFloat(step) / CGFloat(labelCount)
      let price = frame.viewport.priceHigh - (frame.viewport.priceHigh - frame.viewport.priceLow) * Double(ratio)
      let y = frame.mainChartRect.minY + frame.mainChartRect.height * ratio
      let text = formatPrice(price)
      text.draw(
        at: CGPoint(x: priceAxisRect.minX + 6, y: y - frame.theme.axisFont.lineHeight / 2),
        withAttributes: frame.theme.axisAttributes
      )
    }

    // Bybit-style: ticks spaced along visible candle indices; x uses fractional index so labels
    // slide with pan (integer-only xCenter + floor(visibleStart) made text update while x felt stuck).
    let start = frame.viewport.visibleStartIndex
    let end = frame.viewport.visibleEndIndex
    let span = end - start
    let maxSlots = 5
    let slots = min(maxSlots, max(span + 1, 1))
    let denom = max(slots - 1, 1)
    let plotLeft = frame.mainChartRect.minX
    let vs = frame.viewport.scrollOffset
    let sw = frame.viewport.slotWidth
    let count = frame.candleStore.count
    for step in 0..<slots {
      let t = Double(step) / Double(denom)
      let rawIndex = Double(start) + Double(span) * t
      let idx = min(max(Int(rawIndex.rounded()), 0), count - 1)
      let x = plotLeft + CGFloat((rawIndex - vs + 0.5) * sw)
      let timestamp = frame.candleStore.timestamp(at: idx)
      let text = formatTimestamp(timestamp: timestamp, timeframe: frame.candleStore.timeframe)
      let textSize = text.size(withAttributes: frame.theme.axisAttributes)
      let drawX = x - textSize.width / 2
      text.draw(
        at: CGPoint(x: drawX, y: timeAxisRect.minY + 2),
        withAttributes: frame.theme.axisAttributes
      )
    }
  }

  private func formatPrice(_ price: Double) -> NSString {
    NSString(format: "%.2f", price)
  }

  private func formatTimestamp(timestamp: Int64, timeframe: String) -> NSString {
    let formatter = DateFormatter()
    formatter.locale = Locale(identifier: "en_US_POSIX")
    switch timeframe {
    case "d1", "w1", "M1":
      formatter.dateFormat = "MM-dd"
    default:
      formatter.dateFormat = "HH:mm"
    }
    let date = Date(timeIntervalSince1970: TimeInterval(timestamp) / 1000.0)
    return formatter.string(from: date) as NSString
  }
}

File: android/src/main/kotlin/com/tradechart/plugin/renderer/AxisRenderer.kt
package com.tradechart.plugin.renderer

import android.text.format.DateFormat
import com.tradechart.plugin.engine.ChartFrame
import java.util.Date
import kotlin.math.roundToInt

class AxisRenderer : ChartLayerRenderer {
    override fun render(frame: ChartFrame) {
        val priceAxisRect = frame.priceAxisRect
        val timeAxisRect = frame.timeAxisRect
        if (!frame.config.showAxis || priceAxisRect == null || timeAxisRect == null) {
            return
        }

        frame.canvas.drawLine(
            frame.mainChartRect.right,
            frame.mainChartRect.top,
            frame.mainChartRect.right,
            frame.mainChartRect.bottom,
            frame.theme.axisPaint,
        )
        frame.canvas.drawLine(
            frame.mainChartRect.left,
            frame.mainChartRect.bottom,
            frame.mainChartRect.right,
            frame.mainChartRect.bottom,
            frame.theme.axisPaint,
        )

        if (!frame.viewport.hasVisibleRange()) {
            return
        }

        val labelCount = 4
        for (step in 0..labelCount) {
            val ratio = step / labelCount.toFloat()
            val price = frame.viewport.priceHigh - (frame.viewport.priceHigh - frame.viewport.priceLow) * ratio
            val y = frame.mainChartRect.top + frame.mainChartRect.height() * ratio
            frame.canvas.drawText(
                formatPrice(price),
                priceAxisRect.left + 6f,
                y - 4f,
                frame.theme.axisTextPaint,
            )
        }

        val start = frame.viewport.visibleStartIndex
        val end = frame.viewport.visibleEndIndex
        val span = end - start
        val maxSlots = 5
        val slots = minOf(maxSlots, maxOf(span + 1, 1))
        val denom = maxOf(slots - 1, 1)
        val paint = frame.theme.axisTextPaint
        val plotLeft = frame.mainChartRect.left
        val vs = frame.viewport.scrollOffset
        val sw = frame.viewport.slotWidth
        val count = frame.candleStore.count
        for (step in 0 until slots) {
            val t = step.toDouble() / denom.toDouble()
            val rawIndex = start.toDouble() + span.toDouble() * t
            val idx = rawIndex.roundToInt().coerceIn(0, count - 1)
            val x = plotLeft + ((rawIndex - vs + 0.5) * sw).toFloat()
            val timestamp = frame.candleStore.timestampAt(idx)
            val label = formatTimestamp(timestamp, frame.candleStore.timeframe)
            val tw = paint.measureText(label)
            frame.canvas.drawText(
                label,
                x - tw / 2f,
                timeAxisRect.top + paint.textSize + 2f,
                paint,
            )
        }
    }

    private fun formatPrice(price: Double): String = String.format("%.2f", price)

    private fun formatTimestamp(timestamp: Long, timeframe: String): String {
        val pattern = when (timeframe) {
            "d1", "w1", "M1" -> "MM-dd"
            else -> "HH:mm"
        }
        return DateFormat.format(pattern, Date(timestamp)).toString()
    }
}
